File: src/html_document.rb
class HtmlDocument
  attr_reader :html_document, :nodes
  def initialize
    start_nodes = ["<!DOCTYPE html>", "<html>", "<head>", "<meta charset=\"UTF-8\">", "</head>"]
    start_nodes << html_close
    @nodes = start_nodes
    @html_document = @nodes.join("\n")
  end

  def add_node(new_node)
    @nodes.pop # remove html close node
    @nodes << new_node
    @nodes << html_close
  end

  def add_nodes(nodes)
    nodes.each { |node| add_node(node) }
    @html_document = @nodes.join
  end

  def html_close
    "</html>"
  end
end

File: onepg/lib/onepg/html_document.ex
defmodule Onepg.HtmlDocument do
  @base_nodes ["<!DOCTYPE html>", "<html>", "<head>", "<meta charset=\"UTF-8\">", "</head>", "</html>"]

  defstruct [body: Enum.join(@base_nodes, ""), nodes: @base_nodes]

  def add_node(nodes, node) when is_list(nodes) and is_binary(node) do
    {html_close, nodes} = List.pop_at(nodes, -1)
    nodes = nodes ++ [node] ++ [html_close]
    body = Enum.join(nodes, "\n")
    %__MODULE__{body: body, nodes: nodes}
  end

  def add_nodes(existing_nodes, nodes_to_add) when is_list(existing_nodes) and is_list(nodes_to_add) do
    [html_close | existing_tags] = Enum.reverse(existing_nodes)
    tags_without_close = Enum.reverse(existing_tags)
    nodes = tags_without_close ++ nodes_to_add ++ [html_close]
    %__MODULE__{nodes: nodes, body: Enum.join(nodes, "\n")}
  end
end
